rbd: Implement snapshot and clone from snapshot

Madhu-1 opened the pull request:
This Adds support for create,delete snapshot and creating a new rbd image from the snapshot.
    
 ##  Create a snapshot
    
* Create a temporary snapshot from the parent volume
* Clone a new image from a temporary snapshot with options --rbd-default-clone-format 2 --image-feature layering,deep-flatten
* Delete temporary snapshot created
* Create snapshot from clone image
* Check the image chain depth, if the Softlimit is reached Add a task Flatten the cloned image and return success. if the depth is reached hard limit Add a task Flatten the cloned image and return snapshot status ready as false
    
    ```bash
    1) rbd snap create <RBD image for src k8s volume>@<random snap name>
    2) rbd clone --rbd-default-clone-format 2 --image-feature
    layering,deep-flatten <RBD image for src k8s volume>@<random snap>
     <RBD image for temporary snap image>
    3) rbd snap rm <RBD image for src k8s volume>@<random snap name>
    4) rbd snap create <RBD image for temporary snap image>@<RBD image for temporary snap image>
    5) check the depth, if the depth is greater than configured hard
    limit add a task to flatten the cloned image return snapshot status
    ready as false if the depth is greater than soft limit add a task
    to flatten the image and return success
    
    ```
    
## Create a clone from a snapshot
    
* Clone a new image from the snapshot with user-provided options
* Check the depth(n) of the cloned image if n>=(hard limit) Add a task to flatten the image and return ABORT (to avoid image leak)
    
    ```bash
    2) rbd clone --rbd-default-clone-format 2 --image-feature
    <k8s dst vol config> <RBD image for temporary snap image>@<random snap name>
     <RBD image for k8s dst vol>
    4) check the depth, if the depth is greater than the configured hard limit
    add a task to flatten the cloned image return ABORT error if the depth is
    greater than soft limit add a task to flatten the image and return success
    ```
    
## Delete snapshot or PVC
    
* Move the temporary cloned image to the trash
* Add task to remove the image from the trash
    
    ```bash
    1) rbd trash mv <cloned image>
    2) ceph rbd task trash remove <cloned image>
    ```
    
With earlier implementation to delete the image, we used to add a task to remove the image with new changes this cannot be done as the image may contain snapshots or links.
   
    
Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Note: Created a new PR to make review easy and addressed review comments in https://github.com/ceph/ceph-csi/pull/693 and https://github.com/ceph/ceph-csi/pull/467

closes #693 
closes #467 
closes https://github.com/ceph/ceph-csi/issues/263
closes https://github.com/ceph/ceph-csi/issues/446


The snapshot, creating a PVC from Restore and creating a PVC from PVC will be done in different PR's

TODO

- [x] Implement snapshot create, delete and create PVC from a snapshot
- [ ] Create PVC from PVC
- [ ] Fix issues with snapshot/clone (512) limits https://github.com/ceph/ceph-csi/issues/1098 
- [ ] Allow Parallel cloning and snapshots creation on one PVC/clone (Need to work on in-memory locking) 
- [ ] Improve/Add Documentation for snapshot and cloning
- [ ] Check Anything Needed to support datapool
- [ ] Add doc file for the commands
- [ ] If required some code cleanup
- [ ] Replace GetStoredImageID with GetImageAttributes in all places

Madhu-1 commented:
Converted to Draft as scale testing is pending and also need to add TODO for remaining tasks

Madhu-1 commented:
still need to do some smaller changes, marked as ready of review.

Madhu-1 commented:
@dillaman PTAL

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -335,38 +330,62 @@ func deleteImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) er
+	klog.V(4).Infof(util.Log(ctx, "rbd: delete %s using mon %s, pool %s"), image, pOpts.Monitors, pOpts.Pool)
 
-	klog.V(4).Infof(util.Log(ctx, "rbd: rm %s using mon %s"), pOpts, pOpts.Monitors)
+	var ioctx *rados.IOContext
```
just use `pOpts.open()` to get the imahe, and use `pOpts.ioctx` after that for the IOContext. No need to destroy the IOContext, as it will be cleaned up by `pOpts.Destroy()` at a later time.

You need to defer closing the image though.

Madhu-1 replied:
would like to avoid opening an image as its not required it.

nixpanic replied:
ah, right... Calling librbd from outside a `rbdVolume` method is not very nice, but there is currently no cleaner alternative

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -262,7 +294,7 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 }
 
-func createBackingImage(ctx context.Context, cr *util.Credentials, rbdVol *rbdVolume, rbdSnap *rbdSnapshot) error {
+func (cs *ControllerServer) createBackingImage(ctx context.Context, cr *util.Credentials, rbdVol *rbdVolume, rbdSnap *rbdSnapshot) error {
```
why make this a method for ControllerServer? it is not used in the function...

Madhu-1 replied:
This is required for locking

nixpanic replied:
but it is not used in this function? or maybe a follow up commit does that? It looks unneeded in https://github.com/ceph/ceph-csi/pull/1160/commits/b5209f8f74bde7b2a84f287c9b5c58e09e02a751

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -552,6 +554,16 @@ func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials,
 
+	rbdVol.ImageID, err = j.GetReservedImageID(ctx, rbdVol.JournalPool, rbdVol.ReservedID, cr)
+	if err != nil {
+		// skip Not found error for older RBD images, as we have not saved
```
should we insert the image id here? might make it easier for upgrades? file an issue and mark as TODO?

Madhu-1 replied:
it's already handled in deleteVolume, if required later we can move it here also

nixpanic replied:
no idea if it is needed, maybe others can recommend (against) it?

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -773,19 +773,17 @@ func imageFeaturesToUint64(ctx context.Context, imageFeatures string) uint64 {
-
-	output, err := execCommand("rbd", args)
+func createSnapshot(ctx context.Context, vol *rbdVolume, pOpts *rbdSnapshot, cr *util.Credentials) error {
+	klog.V(4).Infof(util.Log(ctx, "rbd: snap create %s/%s@%s using mon %s"), pOpts.Pool, pOpts.RbdImageName, pOpts.RbdSnapName, pOpts.Monitors)
```
just print `%s` <- `pOpts` here instead of building an image name

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -786,24 +786,18 @@ func createSnapshot(ctx context.Context, vol *rbdVolume, pOpts *rbdSnapshot, cr
+	defer image.Close()
 
-	return nil
+	snap := image.GetSnapshot(pOpts.RbdSnapName)
```
please check if `snap` is `nil`

Madhu-1 replied:
no idea in which case it can return `nil` as per https://github.com/ceph/go-ceph/blob/v0.2.0/rbd/rbd.go#L958. adding nil check for safer side.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -786,24 +786,18 @@ func createSnapshot(ctx context.Context, vol *rbdVolume, pOpts *rbdSnapshot, cr
-
-	output, err := execCommand("rbd", args)
+func deleteSnapshot(ctx context.Context, vol *rbdVolume, pOpts *rbdSnapshot, cr *util.Credentials) error {
+	klog.V(4).Infof(util.Log(ctx, "rbd: snap rm %s/%s@%s using mon %s"), pOpts.Pool, pOpts.RbdImageName, pOpts.RbdSnapName, pOpts.Monitors)
```
same here, just use `pOpts`

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -800,18 +803,36 @@ func deleteSnapshot(ctx context.Context, vol *rbdVolume, pOpts *rbdSnapshot, cr
-// nolint: interfacer
-func restoreSnapshot(ctx context.Context, pVolOpts *rbdVolume, pSnapOpts *rbdSnapshot, cr *util.Credentials) error {
-	var output []byte
+func cloneRbdImageFromSnapshot(ctx context.Context, pVolOpts *rbdVolume, pSnapOpts *rbdSnapshot, cr *util.Credentials) error {
```
This is a nice chance on moving `cloneRbdImageFromSnapshot()` to a method of `rbdSnapshot`. Can be done as a cleanup later.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -800,18 +803,36 @@ func deleteSnapshot(ctx context.Context, vol *rbdVolume, pOpts *rbdSnapshot, cr
+		return errors.Wrapf(err, "failed to set image features")
+	}
+
+	ioctx, err := pVolOpts.conn.GetIoctx(pVolOpts.Pool)
```
use `pVolOpts.openIoctx()` and `pVolOpts.ioctx` after that

Madhu-1 replied:
sorry already done, not pushed the latest change :(

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -161,6 +160,7 @@ func buildCreateVolumeResponse(ctx context.Context, req *csi.CreateVolumeRequest
 }
 
 // CreateVolume creates the volume in backend
+// nolint: gocyclo
```
add a comment why `gocyclo` should not be checked here

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -294,27 +259,123 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 	return &csi.CreateVolumeResponse{Volume: volume}, nil
 }
 
+func (cs *ControllerServer) checkFlatten(ctx context.Context, rbdVol *rbdVolume, cr *util.Credentials) error {
```
should this not be a method of `rbdVolume` instead? `cs` is not used in this function, only `rbdVolume`.

Madhu-1 replied:
earlier i stored the hard and soft depth flags in cs. after that i removed. fixed now

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -294,27 +259,123 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
+
+func (cs *ControllerServer) createVolumeFromSnapshot(ctx context.Context, cr *util.Credentials, rbdVol *rbdVolume, snapshotID string) error {
+	rbdSnap := &rbdSnapshot{}
+	if acquired := cs.SnapshotLocks.TryAcquire(rbdSnap.SnapID); !acquired {
```
`rbdSnap.SnapID` is not set? The object just got created with empty values. Maybe you meant to use `snapshotID`?

Madhu-1 replied:
this should be fine, this function will be called after validating the snapshotID. this cannot be unset.

nixpanic replied:
No, I mean you create it on line 283 with `rbdSnap := &rbdSnapshot{}` and then use `rbdSnap.SnapID` on line 284. `rbdSnap.SnapID` will always be an empty string?

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -294,27 +259,123 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
+		return err
+	}
+
+	err = flattenRbdImage(ctx, rbdVol, cr)
```
`flattenRbdImage` should be a method of `rbdVolume`, can be done at a later point too

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -660,33 +734,81 @@ func (cs *ControllerServer) validateSnapshotReq(ctx context.Context, req *csi.Cr
-	// uuid for use as the snap name
+func (cs *ControllerServer) doSnapshot(ctx context.Context, parentVol *rbdVolume, rbdSnap *rbdSnapshot, cr *util.Credentials) (bool, *rbdVolume, error) {
+	// generate cloned volume details from snapshot
+	cloneRbd := generateVolFromSnap(rbdSnap)
```
can this return `nil`?

Madhu-1 replied:
not possible. it will generate the volume struct from snapshot struct(snap struct wont be empty)

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -355,13 +353,12 @@ func deleteImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) er
 
 	var ioctx *rados.IOContext
-	ioctx, err = pOpts.conn.GetIoctx(pOpts.Pool)
+	err = pOpts.openIoctx()
```
hmm, in a follow-up change you call `openIoctx()`. Still easier to call `open()`

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -480,15 +466,12 @@ func genSnapFromSnapID(ctx context.Context, rbdSnap *rbdSnapshot, snapshotID str
 // genVolFromVolID generates a rbdVolume structure from the provided identifier, updating
 // the structure with elements from on-disk image metadata as well
-func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials, secrets map[string]string) (*rbdVolume, error) {
+func genVolFromVolID(ctx context.Context, volumeID string, cleanupVol bool, cr *util.Credentials, secrets map[string]string) (*rbdVolume, error) {
```
I was wondering about the new boolean that was passed to this function. However, `cleanupVol` does not help me at all. Maybe document its intention?

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -568,10 +551,71 @@ func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials,
 	return rbdVol, err
 }
 
+func removeImageFromTrash(rbdVol *rbdVolume) error {
```
just make this a method of `rbdVolume`

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -568,10 +551,71 @@ func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials,
 }
 
+func removeImageFromTrash(rbdVol *rbdVolume) error {
+	ioctx, err := rbdVol.conn.GetIoctx(rbdVol.Pool)
```
call `rbdVol.openIoctx()` and use `rbdVol.ioctx` (no need to destroy the ioctx)

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -568,10 +551,71 @@ func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials,
+	}
+
+	return info, ErrSnapNotFound{rbdSnap.RbdSnapName, fmt.Errorf("snap (%s) for image (%s) not found",
+		rbdSnap.RbdSnapName, rbdSnap.Pool+"/"+rbdSnap.RbdImageName)}
```
use `rbdSnap.String()` from #1117 (or have it automatically done)

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -902,58 +933,99 @@ func (rv *rbdVolume) getImageInfo(ctx context.Context, monitors string, cr *util
-	Name      string `json:"name"`
-	Size      int64  `json:"size"`
-	Timestamp string `json:"timestamp"`
+func getCloneDepth(ctx context.Context, monitors, poolName, imageName string, depth int, cr *util.Credentials) (int, error) {
```
can this be a method of rbdVolume instead? you're passing all important members of the type, and are constructing a new object from those

no need to pass `depth`?

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -902,58 +933,99 @@ func (rv *rbdVolume) getImageInfo(ctx context.Context, monitors string, cr *util
-		"-c", util.CephConfigPath,
-		"--format="+"json",
-		"snap", "ls", rs.imageSpec())
+	depth, err = getCloneDepth(ctx, rbdVol.Monitors, rbdVol.Pool, rbdVol.RbdImageName, 0, cr)
```
why not use `rbdVol` directly in `getCloneDepth()`?

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -902,58 +933,99 @@ func (rv *rbdVolume) getImageInfo(ctx context.Context, monitors string, cr *util
+	}
+	klog.Infof(util.Log(ctx, "clone depth is (%d), configured softlimit (%v) and hardlimit (%v) for %s"), depth, rbdSoftMaxCloneDepth, rbdHardMaxCloneDepth, rbdVol)
+
+	if (depth != 0 && flattenImage) || (depth >= int(rbdHardMaxCloneDepth) || depth >= int(rbdSoftMaxCloneDepth)) {
```
conversion of `rbdHardMaxCloneDepth` seems clutter, can these configuration options not be an int?

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -902,58 +933,99 @@ func (rv *rbdVolume) getImageInfo(ctx context.Context, monitors string, cr *util
+
+// rbdManagerTaskAddFlatten adds a ceph manager task to flatten image
+// asynchronously in the background. If command is not found returns a bool set to false
+func rbdManagerTaskAddFlatten(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) (bool, error) {
```
probably should be a method of `rbdVolume`

nixpanic replied:
this really should return an error `ErrCmdNotFound` or similar instead of an additional `bool`.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -902,58 +933,99 @@ func (rv *rbdVolume) getImageInfo(ctx context.Context, monitors string, cr *util
+	klog.Infof(util.Log(ctx, "clone depth is (%d), configured softlimit (%v) and hardlimit (%v) for %s"), depth, rbdSoftMaxCloneDepth, rbdHardMaxCloneDepth, rbdVol)
+
+	if (depth != 0 && flattenImage) || (depth >= int(rbdHardMaxCloneDepth) || depth >= int(rbdSoftMaxCloneDepth)) {
+		ns, err := rbdManagerTaskAddFlatten(ctx, rbdVol, cr)
```
I guess `ns` stands for `NotSupported`? Better to return an error type that matches that instead of a bool

Madhu-1 replied:
This is not added as part of this PR, if the command is not supported we will kind of fallback.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -902,58 +933,99 @@ func (rv *rbdVolume) getImageInfo(ctx context.Context, monitors string, cr *util
+				if err != nil {
+					return err
+				}
+				rbdImage := librbd.GetImage(rbdVol.ioctx, rbdVol.RbdImageName)
```
use `rbdVol.open()` to get the image

nixpanic commented on internal/rbd/snapshot.go:
```diff
@@ -0,0 +1,124 @@
+		return err
+	}
+
+	cloneFailed := false
```
default to `true`, always assume it fails so that the cleanup path is preferred

Madhu-1 replied:
removed this variable for cleaner approach

nixpanic commented on internal/rbd/snapshot.go:
```diff
@@ -0,0 +1,124 @@
+		klog.Errorf(util.Log(ctx, "failed to delete snapshot: %v"), errSnap)
+	}
+
+	if cloneFailed {
```
not getting cleaned up now, `cloneFailed` is not changed  from `false`

Madhu-1 replied:
removed this variable for cleaner approach

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -191,6 +190,7 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		klog.Errorf(util.Log(ctx, "failed to connect to volume %v: %v"), rbdVol.RbdImageName, err)
 		return nil, status.Error(codes.Internal, err.Error())
 	}
+	defer rbdVol.Destroy()
```
no, I dint get what is the issue. Destroy is called right after the Connect.

nixpanic replied:
yes, but it was called after `parseVolCreateRequest()`, why change that?
See https://github.com/ceph/ceph-csi/pull/1160/commits/33238dccf0831e96374e174840cca60c6c5d194e#diff-f010f466063f62f823361f972d08b554L176-L193

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -299,17 +300,12 @@ func rbdStatus(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) (boo
-	}
+	klog.V(4).Infof(util.Log(ctx, "executing %v for image (%s) using mon %s, pool %s"), args, pOpts.RbdImageName, pOpts.Monitors, pOpts.Pool)
 
 	output, err := execCommand("ceph", args)
```
task manager commands can only be added with "ceph"

nixpanic replied:
is it not **always** `ceph rbd task add ...`, having that as a first constant for args would be better

Madhu-1 commented:
@nixpanic addressed review comments. PTAL. as i need to do follow up PR's to address TODO if any minor changes are required i can do it. 

@dillaman @ShyamsundarR @humblec can you please review this one?

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -345,6 +344,13 @@ func (cs *ControllerServer) DeleteLegacyVolume(ctx context.Context, req *csi.Del
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	defer rbdVol.Destroy()
```
Same here, the defer for `Destroy()` can be done early

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -335,38 +331,62 @@ func deleteImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) er
+		return fmt.Errorf("rbd %s is still being used", image)
 	}
+	// Support deleting the older rbd images whose imageID is not stored in omap
+	if pOpts.ImageID == "" {
```
the `pOpts.ImageID == ""` check should be inside `pOpts.getImageID()`, so that you can always call the function.

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -758,11 +863,61 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	defer rbdVol.Destroy()
```
```defer rbdVol.Destroy()``` can be called immediately after `generateVolFromSnap()`

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -481,15 +467,14 @@ func genSnapFromSnapID(ctx context.Context, rbdSnap *rbdSnapshot, snapshotID str
 // genVolFromVolID generates a rbdVolume structure from the provided identifier, updating
 // the structure with elements from on-disk image metadata as well
-func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials, secrets map[string]string) (*rbdVolume, error) {
+// cleanupVol will be set during the DeleteVolume operation to delete the image,
```
this adds more complexity to the function. is it not easier to understand if the DeleteVolume operation takes care of it? A better maintainable solution has my preference.

Madhu-1 replied:
no other rbdvolume related information is passed from genVolFromVolID to delete the image, in case of failure we return nil and error

nixpanic replied:
hmm, I don't like it. It should probably get cleaned up at one point.

Madhu-1 replied:
This is fixed now.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -768,16 +812,19 @@ func imageFeaturesToUint64(ctx context.Context, imageFeatures string) uint64 {
 	features := uint64(0)
 
 	for _, f := range strings.Split(imageFeatures, ",") {
```
You can use `librbd.FeatureSetFromNames(strings.Split(imageFeatures, ",")` and get the int value directly. Use `Names()` to get a `[]string` with the features that are recognized and see if it misses any of the original requested features.

Also, the strings for features are available like `FeatureNameDeepFlatten`.

Doing it this way, even features that are not known to ceph-csi can get enabled (by advanced users).

Madhu-1 replied:
maybe in followup PR, as I just added one more case  to it.

Madhu-1 commented:
@nixpanic GitHub is not allowing me to reply to a few comments, let me know if we need to discuss any particular comment.

nixpanic commented:
> @nixpanic GitHub is not allowing me to reply to a few comments, let me know if we need to discuss any particular comment.

You might be able to reply if you click the `Commits` tab, the first commit, and then `Next >` until you find a comment you want to respond to? That is also how I left the comments.

phlogistonjohn commented on internal/rbd/controllerserver.go:
```diff
@@ -262,27 +257,121 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
+func (cs *ControllerServer) createBackingImage(ctx context.Context, cr *util.Credentials, rbdVol *rbdVolume, rbdSnap *rbdSnapshot) error {
 	var err error
 
+	var j = new(journal.Connection)
```
Why new? It's exceedingly rare to see a call to new used in most go code IMO.

Madhu-1 replied:
>internal/rbd/controllerserver.go:783:2: SA4006: this value of `j` is never used (staticcheck)
	j := &journal.Connection{}

the short variable declaration was giving a static check error for this now, now change back to address of struct.

ShyamsundarR commented:
>     * we need to have some flag to indicate that all RBD images for PVCs need to be flattened upon creation. This flag would be set to `true` if they have at least one compute / worker node with an old OS that doesn't support the deep-flatten feature and the operations feature that is set when you use clone v2 (setting of the flag can be an exercise for the user / Rook).

From a standalone csph-csi perspective, the NodeStageVolume call can possibly detect the local OS version and also the image depth that is required to be mapped, and decide to allow/disallow the operation.

dillaman commented:
> > ```
> > * we need to have some flag to indicate that all RBD images for PVCs need to be flattened upon creation. This flag would be set to `true` if they have at least one compute / worker node with an old OS that doesn't support the deep-flatten feature and the operations feature that is set when you use clone v2 (setting of the flag can be an exercise for the user / Rook).
> > ```
> 
> From a standalone csph-csi perspective, the NodeStageVolume call can possibly detect the local OS version and also the image depth that is required to be mapped, and decide to allow/disallow the operation.

If we just wanted to randomly fail at that step, you could just let it attempt to `rbd block map` and that will fail as well.

Madhu-1 commented:
> > ```
> > * we need to have some flag to indicate that all RBD images for PVCs need to be flattened upon creation. This flag would be set to `true` if they have at least one compute / worker node with an old OS that doesn't support the deep-flatten feature and the operations feature that is set when you use clone v2 (setting of the flag can be an exercise for the user / Rook).
> > ```
> 
> From a standalone csph-csi perspective, the NodeStageVolume call can possibly detect the local OS version and also the image depth that is required to be mapped, and decide to allow/disallow the operation.

This could give a bad user experience- Any cloned or restored PVC cannot be mapped on the nodes where the kernel version is <8.2+(the user has the manually reschedule the app to the nodes where the kernel is 8.2+ with some affinity). instead of this, we can flatten the image. The only issue with this is Creation operation can be slow. eventually when users move to new kernel versions we can disable the flag by default

humblec commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -107,6 +107,15 @@ provisioner:
+  softMaxCloneDepth: 4
+  # Flatten all cloned rbd images upon creation, if kernel version doesnot
+  # support deep-flatten feature
+  flattenOnCreate: true
```
may `enabledeepflatten` would be a better naming I believe.

Madhu-1 replied:
The name was suggested by @dillaman in one of the discussion, This seems correct as we need to flatten the image upon creation, I don't have hard feelings I can change it if required.

humblec replied:
No, thats fine. lets keep it as `flattenOnCreate`

dillaman replied:
Alternative suggestion could be `enableThinProvisionedClones` (or similar). I don't think anything referencing the RBD `deep-flatten` feature bit would mean much to end-users.

humblec replied:
May be what we do internally - `flatten at create` or `do it in a different queue` also may not add value to the user. in that sense something like `enableThinProvisionedClone` looks better to me.

ShyamsundarR commented:
> > > ```
> > > * we need to have some flag to indicate that all RBD images for PVCs need to be flattened upon creation. This flag would be set to `true` if they have at least one compute / worker node with an old OS that doesn't support the deep-flatten feature and the operations feature that is set when you use clone v2 (setting of the flag can be an exercise for the user / Rook).
> > > ```
> > 
> > 
> > From a standalone csph-csi perspective, the NodeStageVolume call can possibly detect the local OS version and also the image depth that is required to be mapped, and decide to allow/disallow the operation.
> 
> This could give a bad user experience- Any cloned or restored PVC cannot be mapped on the nodes where the kernel version is <8.2+(the user has the manually reschedule the app to the nodes where the kernel is 8.2+ with some affinity). instead of this, we can flatten the image. The only issue with this is Creation operation can be slow. eventually when users move to new kernel versions we can disable the flag by default

I am not suggesting we do away with a flag, I am suggesting that we also add a check on the nodeplugin to detect and prevent the Staging.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -707,115 +842,76 @@ func imageFeaturesToUint64(ctx context.Context, imageFeatures string) uint64 {
 
-	if err != nil {
-		return errors.Wrapf(err, "failed to restore snapshot, command output: %s", string(output))
+	if pVolOpts.ImageFeatures != "" {
```
Note, `ImageFeatures` was removed, use `imageFeatureSet` instead.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -843,62 +939,79 @@ func (rv *rbdVolume) getImageInfo() error {
+			Pool:         poolName,
+			RbdImageName: imageName,
+		}
+		err := vol.getImageInfo()
```
This most likely connects the rbdVolume, so it needs to be destroyed afterwards. Easiest is to always add a `defer vol.Destroy()` immediately after creating the rbdVolume object.

Madhu-1 replied:
will make changes to reuse the ioctx and conn of parent volume, no need to destroy it, as parent will destroy it.

nixpanic replied:
try not to over complicate things, if the connection exists. Optimizations like this can be done later when ioctx's are in a shared pool.

Madhu-1 replied:
we cannot use getImageInfo as it doesn't return the parent Image name, filed an issue https://github.com/ceph/go-ceph/issues/312. in followup PR i will change back to getImageInfo.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -843,62 +939,79 @@ func (rv *rbdVolume) getImageInfo() error {
+			}
+		}
+		if !ns {
+			klog.Errorf(util.Log(ctx, "rbd task add flatten not supported will flatten image once hardlimit is reached: %v"), err)
```
maybe quote the command, or describe it like "task manager does not support flatten, image will be .."

nixpanic commented on internal/rbd/snapshot.go:
```diff
@@ -0,0 +1,122 @@
+	"context"
+	"strings"
+
+	"github.com/ceph/ceph-csi/internal/util"
```
I think it is recommended to place imports of the local project after all others

Madhu-1 replied:
Nope i sent a PR to change order as per the standards.

nixpanic replied:
oh, ok, my mistake

nixpanic commented on internal/rbd/snapshot.go:
```diff
@@ -0,0 +1,122 @@
+	"k8s.io/klog"
+)
+
+func createRBDClone(ctx context.Context, parentVol, cloneRbdVol *rbdVolume, snap *rbdSnapshot, cr *util.Credentials) error {
```
Please consider making all/most of the functions in this file methods of the rbdSnapshot type

Madhu-1 replied:
will take care of it when doing fixing the TODO.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -707,115 +842,76 @@ func imageFeaturesToUint64(ctx context.Context, imageFeatures string) uint64 {
 
 	for _, f := range strings.Split(imageFeatures, ",") {
-		if f == "layering" {
+		switch f {
```
waiting for #1178 to be merged

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -821,62 +921,79 @@ func (rv *rbdVolume) getImageInfo() error {
+		}
+		err := vol.updateVolWithImageInfo(ctx, rv.conn.Creds)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to check depth on image:%s %s"), vol, err)
```
nit: misplaced colon: `failed to ... depth on %s: %s`

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -821,62 +921,79 @@ func (rv *rbdVolume) getImageInfo() error {
-		return snpInfo, fmt.Errorf("unmarshal failed: %+v. raw buffer response: %s", err, string(stdout))
+		return err
 	}
+	klog.Infof(util.Log(ctx, "clone depth is (%d), configured softlimit (%v) and hardlimit (%v) for %s"), depth, rbdSoftMaxCloneDepth, rbdHardMaxCloneDepth, rv)
```
nit: mixing `%d` and `%v` in the formatting string?

chenxu1990 commented on internal/rbd/controllerserver.go:
```diff
@@ -622,49 +731,81 @@ func (cs *ControllerServer) validateSnapshotReq(ctx context.Context, req *csi.Cr
-	err = createSnapshot(ctx, rbdSnap, cr)
-	// If snap creation fails, even due to snapname already used, fail, next attempt will get a new
-	// uuid for use as the snap name
+func (cs *ControllerServer) doSnapshot(ctx context.Context, parentVol *rbdVolume, rbdSnap *rbdSnapshot, cr *util.Credentials) (bool, *rbdVolume, error) {
```
Is it better to call 'doSnapClone' ?

chenxu1990 commented on internal/rbd/rbd_journal.go:
```diff
@@ -137,26 +140,42 @@ func checkSnapExists(ctx context.Context, rbdSnap *rbdSnapshot, cr *util.Credent
+	if err != nil {
+		return false, err
+	}
+	vol.Destroy()
```
defer ?

ShyamsundarR commented on internal/rbd/snapshot.go:
```diff
@@ -0,0 +1,122 @@
+	errSnap := deleteSnapshot(ctx, parentVol, snap)
+	if errSnap != nil {
+		klog.Errorf(util.Log(ctx, "failed to delete snapshot: %v"), errSnap)
+		delErr := deleteImage(ctx, cloneRbdVol, cr)
```
If there was an `err` in line 37 above, then deleteImage need not be invoked.

Possibly cleaner to defer `deleteSnapshot` post `createSnapshot` and let the rest of the code play out. Similarly wrap an error routine for defering `deleteImage` (latter is just a thought, former seems more useful)

ShyamsundarR commented on internal/rbd/controllerserver.go:
```diff
@@ -622,49 +731,81 @@ func (cs *ControllerServer) validateSnapshotReq(ctx context.Context, req *csi.Cr
+	}
+	defer j.Destroy()
+
+	err = j.ReserveImageID(ctx, rbdSnap.JournalPool, rbdSnap.ReservedID, cloneRbd.ImageID, cr)
```
If we fail just prior to this call (as in provisioner was restarted, say), we would fail to save the image ID. The snapshot although is complete, so on a `CreateSnapshot` retry we would succeed in `checkSnapCloneExists` and not save the image ID for `DeleteSnapshot` call.

Either `checkSnapCloneExists` should ensure the image ID is saved, or undo the entire work, or the image ID should be saved sooner in the transaction to prevent such cases.

ShyamsundarR commented on internal/rbd/controllerserver.go:
```diff
@@ -622,49 +731,81 @@ func (cs *ControllerServer) validateSnapshotReq(ctx context.Context, req *csi.Cr
 	}()
-	err = protectSnapshot(ctx, rbdSnap, cr)
+
+	err = createSnapshot(ctx, cloneRbd, rbdSnap)
```
I am slightly confused by this additional snapshot creation here. Is this the single RBD snapshot from which we will create future clones? As commented [here](https://github.com/ceph/ceph-csi/pull/1160#issuecomment-642069930)?

If so, we again do not seem to be checking if this snapshot is created when we call `checkSnapCloneExists`, as a result we would fail in the future to use this snapshot once it is marked ready?

ShyamsundarR commented on internal/rbd/rbd_journal.go:
```diff
@@ -137,26 +140,42 @@ func checkSnapExists(ctx context.Context, rbdSnap *rbdSnapshot, cr *util.Credent
+	}
+	vol.Destroy()
+
+	err = vol.flattenRbdImage(ctx, cr)
```
(nit) Flatten does not seem to belong to this function as such, which checks if the journal and hence the image is intact. I would see if we can move the flatten check outside of this code to the caller.

ShyamsundarR commented on internal/rbd/controllerserver.go:
```diff
@@ -262,27 +258,121 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 	}
 
-	klog.V(4).Infof(util.Log(ctx, "created volume %s backed by image %s"), rbdVol.RequestName, rbdVol.RbdImageName)
+	err = j.ReserveImageID(ctx, rbdVol.JournalPool, rbdVol.ReservedID, rbdVol.ImageID, cr)
```
We may need to move this further up, like for snapshots, and ensure when we find an existing image, this data is already stored.

ShyamsundarR commented on internal/journal/voljournal.go:
```diff
@@ -649,6 +655,24 @@ func (conn *Connection) GetImageAttributes(ctx context.Context, pool, objectUUID
+}
+
+// GetReservedImageID retrives the reserved image ID from the omap
+func (conn *Connection) GetReservedImageID(ctx context.Context, pool, reservedUUID string, cr *util.Credentials) (string, error) {
```
Ideally we should be able to get this from `GetImageAttributes` (maybe once we have the fetch all keys in a single call).

ShyamsundarR replied:
Tracked in the TODO list.

ShyamsundarR commented on internal/rbd/snapshot.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+func undoSnapshotCloning(ctx context.Context, parentVol *rbdVolume, rbdSnap *rbdSnapshot, cloneVol *rbdVolume, cr *util.Credentials, deleteVol bool) error {
+	err := deleteSnapshot(ctx, parentVol, rbdSnap)
```
This is the snapshot of the cloned image, that would be used for future CSI-cloning, right? Attempting to get my bearing straight on this one :)

Madhu-1 commented:
As per the latest discussion to handle mounting of rbd images or clone images in the chain which has the deep-flatten feature on the nodes who is having lower kernel version will flatten the image before use,
There are two problems we need to address

* We don't have a specific error message to return in NodeStageRequest(intermediate error) currently we are return InternalError to retry with backoff( maybe we need to file an issue in CSI spec to extend node stage error message).as the kubelet retry's when an internal server error occurs we are good here.
* The secret we are using for node stage in the rook doesn't have the permission to talk to the manager to issue rbd flatten images. we need to update rook to extend user caps

Madhu-1 commented:
@ShyamsundarR PTAL.

Madhu-1 commented:
some cleanup required in commits will do it once changes look good.

Madhu-1 commented on cmd/cephcsi.go:
```diff
@@ -77,6 +77,12 @@ func init() {
+	flag.UintVar(&conf.RbdHardMaxCloneDepth, "rbdhardmaxclonedepth", 8, "Hard limit for maximum number of nested volume clones that are taken before a flatten occurs")
+	flag.UintVar(&conf.RbdSoftMaxCloneDepth, "rbdsoftmaxclonedepth", 4, "Soft limit for maximum number of nested volume clones that are taken before a flatten occurs")
+
+	flag.BoolVar(&conf.ForceFlattenImage, "forceflatten", true,
```
Need to flip this value and change the logic to flatten in nodestage

Madhu-1 commented:
@ShyamsundarR @nixpanic @humblec PTAL

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -60,6 +61,22 @@ type stageTransaction struct {
 }
 
+var (
+	release = ""
```
call this `kernelRelease` or `kernelVersion` please. Just `release` is confusing (was it attached?).

nixpanic commented on internal/util/util.go:
```diff
@@ -94,9 +95,18 @@ type Config struct {
 	Version            bool // cephcsi version
 
+	// SkipForceFlatten is set to false if the kernel supports mounting of
+	// rbd image/image chain has the deep-flatten feature.
```
instead of `image/image`, maybe write `image or the image` to be clearer

nixpanic commented on internal/util/util.go:
```diff
@@ -151,6 +161,90 @@ func KernelVersion() (string, error) {
 	return strings.TrimRight(string(utsname.Release[:]), "\x00"), nil
 }
 
+// KernelVersion holds kernel related informations
```
This could be it's own PR that can get reviewed+merged quicker

Madhu-1 replied:
@nixpanic as no other PR is dependent on this one, as it's in a separate commit lets get it done here only.

nixpanic replied:
sure, thats fine

nixpanic commented on internal/util/util.go:
```diff
@@ -143,7 +143,7 @@ func ValidateDriverName(driverName string) error {
 
 // KernelVersion returns the version of the running Unix (like) system from the
```
Need to update the doc with the new function name too

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -306,74 +308,109 @@ func rbdStatus(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) (boo
+// addRbdManagerTask adds a ceph manager task to execute command
+// asynchronously. If command is not found returns a bool set to false
+// example arg ["trash", "remove","pool/image"]
+func addRbdManagerTask(ctx context.Context, pOpts *rbdVolume, arg []string) (bool, error) {
```
I would really like to see a single `error` returned, not `(bool, error)`. It can be `ErrNotSupported` instead of the potential `false`.

Madhu-1 replied:
This is not new implementation the code already , if any changes are required we can do it as an enhancement

nixpanic replied:
ok!

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -175,11 +175,6 @@ func (rs *rbdSnapshot) String() string {
 	return fmt.Sprintf("%s/%s@%s", rs.Pool, rs.RbdImageName, rs.RbdSnapName)
 }
 
-// imageSpec returns the image-spec (pool/image) format of the snapshot
```
Why does commit 62aadd543269f07ab8d934274fc2d62b64259eb3 remove `imageSpec()`? It was not used in the rest of the change of the commit.

Madhu-1 replied:
Imagespec is not used anymore, The relevant changes are removed in final commit you want this change to be part of same commit?

Madhu-1 replied:
corrected

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -757,18 +757,33 @@ func (rv *rbdVolume) deleteSnapshot(ctx context.Context, pOpts *rbdSnapshot) err
-	klog.V(4).Infof(util.Log(ctx, "rbd: clone %s using mon %s"), pVolOpts, pVolOpts.Monitors)
-	args := []string{"clone", pSnapOpts.String(), pVolOpts.String(),
-		"--id", cr.ID, "-m", pVolOpts.Monitors, "--keyfile=" + cr.KeyFile}
+	options := librbd.NewRbdImageOptions()
```
please add `defer options.Destroy()` to prevent memory leaks

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -192,53 +192,49 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 	if found {
+		if rbdSnap != nil {
```
Add some comments why `checkFlatten()` is done? I assume the function may also trigger flattening in case it is needed. Writing it explicitly is very helpful.

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -192,53 +192,49 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
-				return nil, deleteErr
-			}
-			return nil, err
+		if _, ok := err.(ErrFlattenInProgress); ok {
```
use `error.Is(err, ErrFlattenInProgress)` as described at https://github.com/golang/go/wiki/ErrorValueFAQ#how-should-i-change-my-error-handling-code-to-work-with-the-new-features

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -262,27 +258,117 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
+func checkFlatten(ctx context.Context, rbdVol *rbdVolume, cr *util.Credentials) error {
+	err := rbdVol.flattenRbdImage(ctx, cr, false)
+	if err != nil {
+		if _, ok := err.(ErrFlattenInProgress); ok {
```
Same here, use the new `errors.Is()`

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -262,27 +258,117 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
+
+	err := genSnapFromSnapID(ctx, rbdSnap, snapshotID, cr)
+	if err != nil {
+		if _, ok := err.(util.ErrPoolNotFound); ok {
```
Same here, `errors.Is()`

Madhu-1 replied:
Yes I already tried it wont work, would like to keep it as it is

nixpanic replied:
Ok, I'm considering to clean the other occasions up. Since Golang 1.13 these new ways are recommended.

Not a blocker, it can be done later.

Madhu-1 commented:
> The commit message of [db5d142](https://github.com/ceph/ceph-csi/commit/db5d142aa35d9919de1d352cee0430ebf0913c17) (rbd: Implement snapshot and clone from snapshot) contains lots of details. Is this also available in some .md file in the repo? If not, adding it as a design would be great (and mention in the comments in the implementation that the doc exists, so it gets updated)

I can do it as a follow up pr as I still need to address few todo. Will add a md file for both snapshot , restore and clone

Madhu-1 commented:
@nixpanic comments are addressed PTAL

nixpanic commented:
Please keep the PR text updated with TODOs. Add them to the list, and mark them done (don't remove them, keep them listed).

Madhu-1 commented:
> Please keep the PR text updated with TODOs. Add them to the list, and mark them done (don't remove them, keep them listed).

@nixpanic done.

Madhu-1 commented:
created a Tracker issue for TODO https://github.com/ceph/ceph-csi/issues/1188. once this is merged the next plan is to fix the TODO which is present in tracker

Madhu-1 commented:
> I'm ok with getting this merged. There are many TODO's left. What is the plan to address them? If you want to sent them as follow-up PRs that is fine, but each TODO needs an issue in that case, and add it to the open TODO list in this PR.
> 
> Make sure all of the TODO-issues get added to the "Backup & Restore" project, and/or the v3.0.0 milestone. Without that, tracking the status of the feature will be difficult.

opened a single issue to track all the TODO https://github.com/ceph/ceph-csi/issues/1188

Madhu-1 commented:
@ShyamsundarR @nixpanic PTAL.

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -226,6 +245,27 @@ func (ns *NodeServer) stageTransaction(ctx context.Context, req *csi.NodeStageVo
+		}
+	}
+	if !util.CheckKernelSupport(kernelRelease, deepFlattenSupport) {
+		if !skipForceFlatten {
```
you might want to combine these two if-statements so that the code block below is not indented 2x

Madhu-1 replied:
Any suggestion?

Madhu-1 commented:
> Was there a request for `skipForceFlatten`? It would need an e2e test-case, I think.

This cannot be tested in CI as kernel in CI doesnt support deep-flatten
 
> Not sure the commit message `high performance` is suitable. Volume creation might be faster when flattening is skipped, but that also means there is more indirection when reading/writing.

Updated the commit message

dillaman commented on docs/deploy-rbd.md:
```diff
@@ -26,22 +26,25 @@ make image-cephcsi
+| `--domainlabels`         | _empty_               | Kubernetes node labels to use as CSI domain labels for topology aware provisioning, should be a comma separated value (ex:= "failure-domain/region,failure-domain/zone")                                                                                                             |
+| `--rbdhardmaxclonedepth` | `8`                   | Hard limit for maximum number of nested volume clones that are taken before a flatten occurs                                                                                                                                                                                         |
+| `--rbdsoftmaxclonedepth` | `4`                   | Soft limit for maximum number of nested volume clones that are taken before a flatten occurs                                                                                                                                                                                         |
+| `--skipforceflatten`     | `false`               | skip image flattening on kernel < 5.2 which support mapping of rbd images which has the deep-flatten feature                                                                                                                                                                         |
```
I thought we were killing this in favor of just force-flattening an image upon first use on an incompatible node?

Madhu-1 replied:
@dillaman This is just an option provided to the user if the kernel version which supports deep-flatten is not in the list that cephcsi maintains, the user can still skip flattening of the rbd images before use in NodeStageRequest, the user doesn't need to wait for a new cephcsi release which adds the kernel to the list of the kernel it maintains that supports the deep-flatten feature.

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -519,6 +519,11 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
+			err = rbdVol.removeImageFromTrash()
+			if err != nil {
+				klog.Errorf(util.Log(ctx, "failed to delete rbd image: %s from trash with error: %v"), rbdVol, err)
+				return nil, status.Error(codes.Internal, err.Error())
```
Is this a critical error? Is the trash not cleaned up at a later time automatically?
Deleting an image should probably succeed when the image is gone (which could be in the trash).
I think logging the error (and mentioning the trash is cleared later) is sufficient?

Madhu-1 replied:
Removed now, lets ceph handle it, if required this can be added later also

humblec commented on docs/deploy-rbd.md:
```diff
@@ -26,22 +26,25 @@ make image-cephcsi
+| `--timeout`              | `"3s"`                | Probe timeout in seconds                                                                                                                                                                                                                                                             |
+| `--histogramoption`      | `0.5,2,6`             | [Deprecated] Histogram option for grpc metrics, should be comma separated value (ex:= "0.5,2,6" where start=0.5 factor=2, count=6)                                                                                                                                                   |
+| `--domainlabels`         | _empty_               | Kubernetes node labels to use as CSI domain labels for topology aware provisioning, should be a comma separated value (ex:= "failure-domain/region,failure-domain/zone")                                                                                                             |
+| `--rbdhardmaxclonedepth` | `8`                   | Hard limit for maximum number of nested volume clones that are taken before a flatten occurs                                                                                                                                                                                         |
```
@Madhu-1 can we restrict the readme changes only for newly added parameters ?

Madhu-1 replied:
The table alignment need to be changes as the number of char in line increases

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -894,6 +899,11 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
 			klog.Errorf(util.Log(ctx, "failed to delete rbd image: %s/%s with error: %v"), rbdVol.Pool, rbdVol.VolName, err)
 			return nil, status.Error(codes.Internal, err.Error())
 		}
+		err = rbdVol.removeImageFromTrash()
```
Same here, removing from trash is probably not a real failure (from the CSI specs point of view)

humblec commented on e2e/rbd.go:
```diff
@@ -263,33 +263,90 @@ var _ = Describe("RBD", func() {
+					imageList := listRBDImages(f)
+					// total images in cluster is 1 parent rbd image+ total snaps
+					if len(imageList) != totalCount+1 {
+						e2elog.Logf("backend images not matching kubernetes pvc,snap count,image count %d kubernetes resource count %d", len(imageList), totalCount+1)
```
this error message has to be revisited or can be better formatted.

Madhu-1 replied:
Sure, can do a enhancement

humblec commented on e2e/rbd.go:
```diff
@@ -263,33 +263,90 @@ var _ = Describe("RBD", func() {
+						name := fmt.Sprintf("%s%d", f.UniqueName, i)
+						pvcClone.Spec.DataSource.Name = name
+						pvcClone.Namespace = f.UniqueName
+						appClone.Namespace = f.UniqueName
```
these  2 assignments can be clubbed to single.

nixpanic commented on internal/rbd/rbd_journal.go:
```diff
@@ -280,6 +280,24 @@ func (rv *rbdVolume) Exists(ctx context.Context) (bool, error) {
 	}
 
+	rv.ImageID, err = j.GetStoredImageID(ctx, rv.JournalPool, rv.ReservedID, rv.conn.Creds)
+	if _, ok := err.(util.ErrKeyNotFound); ok {
```
is this safe if `err == nil`?

Madhu-1 replied:
Yes its safe

humblec commented on e2e/rbd.go:
```diff
@@ -263,33 +263,90 @@ var _ = Describe("RBD", func() {
+					imageList = listRBDImages(f)
+					// total images in cluster is 1 parent rbd image+ total
+					// snaps+ total clones
+					if len(imageList) != totalCount+totalCount+1 {
```
For clarity , we could make use of `totalSnapCount` or `totalCloneCount` variables instead of relying on `totalCount`

humblec replied:
@Madhu-1 just to confirm these changes are tracked for next set of PRs ?

Madhu-1 replied:
@humblec changes are already done

humblec commented on e2e/rbd.go:
```diff
@@ -263,33 +263,90 @@ var _ = Describe("RBD", func() {
+
+					imageList = listRBDImages(f)
+					// total images in cluster is total snaps+ total clones
+					if len(imageList) != totalCount+totalCount {
```
same here , totalSnapCount for snapshot is better.

Madhu-1 commented:
@humblec @nixpanic can you please review it?
